File: src/main/java/dev/cdh/affiliate/AnimationState.java
package dev.cdh.affiliate;

public final class AnimationState {
    /**
     * 1 ~ 8: frameNum
     * 9 ~ 16: animationSteps
     * 17 ~ 24: bubbleFrame
     * 25 ~ 32: bubbleSteps
     */
    private int packedValue = 0;

    public int frameNum() {
        return packedValue & 0xff;
    }

    public void setFrameNum(int frameNum) {
        packedValue = (packedValue & 0xffffff00) | (frameNum & 0xff);
    }

    public void resetFrame() {
        setFrameNum(0);
    }

    public int animationSteps() {
        return (packedValue >> 8) & 0xff;
    }

    public void setAnimationSteps(int steps) {
        packedValue = (packedValue & 0xffff00ff) | ((steps & 0xff) << 8);
    }

    public void incrementAnimationSteps() {
        setAnimationSteps(animationSteps() + 1);
    }

    public int bubbleFrame() {
        return (packedValue >> 16) & 0xff;
    }

    public void setBubbleFrame(int frame) {
        packedValue = (packedValue & 0xff00ffff) | ((frame & 0xff) << 16);
    }

    public void nextBubbleFrame() {
        setBubbleFrame(bubbleFrame() + 1);
        setBubbleSteps(0);
    }

    public void resetBubbleFrame() {
        setBubbleFrame(0);
    }

    public int bubbleSteps() {
        return (packedValue >> 24) & 0xff;
    }

    public void setBubbleSteps(int steps) {
        packedValue = (packedValue & 0x00ffffff) | ((steps & 0xff) << 24);
    }

    public void incrementBubbleSteps() {
        setBubbleSteps(bubbleSteps() + 1);
    }

    public void reset() {
        packedValue = 0;
    }

    public void nextFrame() {
        setFrameNum(frameNum() + 1);
        setAnimationSteps(0);
    }
}

File: src/main/kotlin/affiliate/AnimationState.kt
package dev.cdh.affiliate

class AnimationState {
    private var frameNum = 0
    private var animationSteps = 0
    private var bubbleFrame = 0
    private var bubbleSteps = 0

    fun frameNum(): Int {
        return frameNum
    }

    fun setFrameNum(frameNum: Int) {
        this.frameNum = frameNum
    }

    fun resetFrame() {
        frameNum = 0
    }

    fun animationSteps(): Int {
        return animationSteps
    }

    fun setAnimationSteps(animationSteps: Int) {
        this.animationSteps = animationSteps
    }

    fun incrementAnimationSteps() {
        animationSteps++
    }

    fun bubbleFrame(): Int {
        return bubbleFrame
    }

    fun setBubbleFrame(bubbleFrame: Int) {
        this.bubbleFrame = bubbleFrame
    }

    fun resetBubbleFrame() {
        bubbleFrame = 0
    }

    fun bubbleSteps(): Int {
        return bubbleSteps
    }

    fun setBubbleSteps(bubbleSteps: Int) {
        this.bubbleSteps = bubbleSteps
    }

    fun incrementBubbleSteps() {
        bubbleSteps++
    }

    fun nextFrame() {
        frameNum++
        animationSteps = 0
    }

    fun nextBubbleFrame() {
        bubbleFrame++
        bubbleSteps = 0
    }

    fun reset() {
        frameNum = 0
        animationSteps = 0
        bubbleFrame = 0
        bubbleSteps = 0
    }
}